symbolic-base/test/Tests/Symbolic/Data/Common.hs

```diff
@@ -0,0 +1,234 @@
+  , Layout (x (Interpreter a)) ~ i
+  , SymbolicOutput (y (Interpreter a))
+  , Context (y (Interpreter a)) ~ Interpreter a
+  , Layout (y (Interpreter a)) ~ o
+  , Payload (y (Interpreter a)) ~ U1
+  ) => ArithmeticCircuit a i o -> x (Interpreter a) -> y (Interpreter a)
+evalCircuit1 ac input =
+  restore $ const (Interpreter $ eval ac $ runInterpreter $ arithmetize input Proxy, U1)
+
+evalCircuit2 :: forall a x y z ix iy i o . (Arithmetic a, Binary a, Representable i, Representable o
+  , SymbolicInput (x (Interpreter a))
+  , Context (x (Interpreter a)) ~ Interpreter a
+  , Layout (x (Interpreter a)) ~ ix
+  , SymbolicInput (y (Interpreter a))
+  , Context (y (Interpreter a)) ~ Interpreter a
+  , Layout (y (Interpreter a)) ~ iy
+  , SymbolicOutput (z (Interpreter a))
+  , Context (z (Interpreter a)) ~ Interpreter a
+  , Layout (z (Interpreter a)) ~ o
+  , Payload (z (Interpreter a)) ~ U1
+  , i ~ ix :*: iy
+  ) => ArithmeticCircuit a i o -> x (Interpreter a) -> y (Interpreter a) -> z (Interpreter a)
+evalCircuit2 ac x y =
+  restore $ const (Interpreter $ eval ac input, U1)
+  where
+    ix = runInterpreter $ arithmetize x Proxy
+    iy = runInterpreter $ arithmetize y Proxy
+    input = ix :*: iy
+
+compileCircuit0 :: forall a x o . (Binary a
+  , SymbolicInput (x (ArithmeticCircuit a U1))
+  , Context (x (ArithmeticCircuit a U1)) ~ ArithmeticCircuit a U1
+  , Layout (x (ArithmeticCircuit a U1)) ~ o
+  , Payload (x (ArithmeticCircuit a U1)) ~ U1
+  ) => (forall c . Symbolic c => x c) -> ArithmeticCircuit a U1 o
+compileCircuit0 v =
+  compileWith @a id (\U1 -> (U1, U1)) $ v @(ArithmeticCircuit a U1)
+
+compileCircuit1 :: forall a x y i o . (Binary a, Representable i
+  , SymbolicInput (x (ArithmeticCircuit a i))
+  , Context (x (ArithmeticCircuit a i)) ~ ArithmeticCircuit a i
+  , Layout (x (ArithmeticCircuit a i)) ~ i
+  , Payload (x (ArithmeticCircuit a i)) ~ U1
+  , SymbolicOutput (y (ArithmeticCircuit a i))
+  , Context (y (ArithmeticCircuit a i)) ~ ArithmeticCircuit a i
+  , Layout (y (ArithmeticCircuit a i)) ~ o
+  ) => (forall c . Symbolic c => x c -> y c) -> ArithmeticCircuit a i o
+compileCircuit1 func =
+  compileWith @a id (\i -> (U1 :*: U1, i :*: U1)) $ func @(ArithmeticCircuit a i)
+
+compileCircuit2 :: forall a x y z ix iy i o . (Binary a, Representable i
+  , SymbolicInput (x (ArithmeticCircuit a i))
+  , Context (x (ArithmeticCircuit a i)) ~ ArithmeticCircuit a i
+  , Layout (x (ArithmeticCircuit a i)) ~ ix
+  , Payload (x (ArithmeticCircuit a i)) ~ U1
+  , SymbolicInput (y (ArithmeticCircuit a i))
+  , Context (y (ArithmeticCircuit a i)) ~ ArithmeticCircuit a i
+  , Layout (y (ArithmeticCircuit a i)) ~ iy
+  , Payload (y (ArithmeticCircuit a i)) ~ U1
+  , SymbolicOutput (z (ArithmeticCircuit a i))
+  , Context (z (ArithmeticCircuit a i)) ~ ArithmeticCircuit a i
+  , Layout (z (ArithmeticCircuit a i)) ~ o
+  , i ~ ix :*: iy
+  ) => (forall c . Symbolic c => x c -> y c -> z c) -> ArithmeticCircuit a i o
```

Comment:
If we specialize `c` with `ArithmeticCircuit a i` here, we can 1) get rid of explicit specialization (`func @(...)`) next 2) generalize `z (ArithmeticCircuit a i)` and the like away